crypto/x509/x509_vfy.c

```diff
@@ -468,46 +468,59 @@ static int check_chain(X509_STORE_CTX *ctx)
+         * and various other checks actually have no (useful) value
+         */
+        CB_FAIL_IF((x->ex_flags & EXFLAG_INVALID) != 0,
+                   ctx, x, i, X509_V_ERR_INVALID_CERTIFICATE_FORMAT);
+
         ret = X509_check_ca(x);
         switch (must_be_ca) {
-        case -1:
+        case -1: /*  we accept both CA and non-CA certificates */
             CB_FAIL_IF((ctx->param->flags & X509_V_FLAG_X509_STRICT) != 0
-                           && ret != 1 && ret != 0,
+                           && ret != X509_CA_TYPE_V3
+                           && ret != X509_CA_TYPE_NON_CA_OR_ERROR,
                        ctx, x, i, X509_V_ERR_INVALID_CA);
             break;
-        case 0:
-            CB_FAIL_IF(ret != 0, ctx, x, i, X509_V_ERR_INVALID_NON_CA);
+        case 0: /* we only accept non-CA certificates */
+            CB_FAIL_IF(ret != X509_CA_TYPE_NON_CA_OR_ERROR,
+                       ctx, x, i, X509_V_ERR_INVALID_NON_CA);
             break;
-        default:
+        default: /* we only accept CA certificates */
             /* X509_V_FLAG_X509_STRICT is implicit for intermediate CAs */
-            CB_FAIL_IF(ret == 0
+            CB_FAIL_IF(ret == X509_CA_TYPE_NON_CA_OR_ERROR
                        || ((i + 1 < num
                             || ctx->param->flags & X509_V_FLAG_X509_STRICT)
-                           && ret != 1), ctx, x, i, X509_V_ERR_INVALID_CA);
+                           && ret != X509_CA_TYPE_V3),
+                       ctx, x, i, X509_V_ERR_INVALID_CA);
             break;
         }
-        if (num > 1) {
-            /* Check for presence of explicit elliptic curve parameters */
-            ret = check_curve(x);
-            CB_FAIL_IF(ret < 0, ctx, x, i, X509_V_ERR_UNSPECIFIED);
-            CB_FAIL_IF(ret == 0, ctx, x, i, X509_V_ERR_EC_KEY_EXPLICIT_PARAMS);
-        }
-        /*
-         * Do the following set of checks only if strict checking is requested
-         * and not for self-issued (including self-signed) EE (non-CA) certs
-         * because RFC 5280 does not apply to them according RFC 6818 section 2.
-         */
-        if ((ctx->param->flags & X509_V_FLAG_X509_STRICT) != 0
-            && num > 1) { /*
-                           * this should imply
-                           * !(i == 0 && (x->ex_flags & EXFLAG_CA) == 0
-                           *          && (x->ex_flags & EXFLAG_SI) != 0)
-                           */
+
+        /* Reject unknown critical extensions due to RFC 5280 section 4.2 */
+        CB_FAIL_IF((ctx->param->flags & X509_V_FLAG_IGNORE_CRITICAL) == 0
+                   && (x->ex_flags & EXFLAG_CRITICAL) != 0,
+                   ctx, x, i, X509_V_ERR_UNHANDLED_CRITICAL_EXTENSION);
+
+        /* Check for presence of explicit elliptic curve parameters */
+        ret = check_curve(x);
+        CB_FAIL_IF(ret < 0, ctx, x, i, X509_V_ERR_UNSPECIFIED);
+        CB_FAIL_IF(ret == 0, ctx, x, i, X509_V_ERR_EC_KEY_EXPLICIT_PARAMS);
```

Comment:
See above.  The story with curve parameter enforcement looks rather murky to me.

Reply:
To me, too.